File: src/snowflake.rs
use std::time::{SystemTime, UNIX_EPOCH};

// A snowflake               1420070400000
pub const VIVA_EPOCH: u128 = 1660769388000;

pub const WORKER_BITS: u8 = 5;
pub const PROCESS_BITS: u8 = 5;
pub const INCREMENT_BITS: u8 = 12;
pub const MAX_WORKER_ID: u32 = 1 << WORKER_BITS;
pub const MAX_PROCESS_ID: u32 = 1 << PROCESS_BITS;
pub const MAX_INCREMENT_ID: u32 = 1 << INCREMENT_BITS;
pub const MAX_TIMESTAMP: u64 = 1 << (64 - WORKER_BITS - PROCESS_BITS - INCREMENT_BITS);


#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Snowflake {
    pub id: u64,
    timestamp: u64,
    worker_id: u64,
    process_id: u64,
    increment: u64,
}

impl Snowflake {
    pub fn new(id: u64) -> Self {
        Self::from_id(id)
    }

    pub fn init_from(id: u64, timestamp: u64, worker_id: u64, process_id: u64, increment: u64) -> Self {
        Self {
            id,
            timestamp,
            worker_id,
            process_id,
            increment,
        }
    }

    pub fn get_timestamp_at(&self) -> u64 {
        ((self.timestamp as u128) + VIVA_EPOCH) as u64
    }

    /// A method to get the current timestamp.
    pub fn get_timestamp(&self) -> u64 {
        self.timestamp
    }

    /// A method to get the worker id.
    pub fn get_worker_id(&self) -> u64 {
        self.worker_id
    }

    /// A method to get the process id.
    pub fn get_process_id(&self) -> u64 {
        self.process_id
    }

    /// A method to get the increment.
    pub fn get_increment(&self) -> u64 {
        self.increment
    }

    pub fn get_id(&self) -> u64 {
        self.id
    }


    /// This function will create a Snowflake from the current time, the worker ID, the process ID, and an increment.
    /// The increment is used to ensure that IDs are unique across a worker process.
    pub fn create(worker_id: u64, process_id: u64, increment: u64) -> Self {
        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos();
        let timestamp: u64 = ((now - VIVA_EPOCH) % (MAX_TIMESTAMP as u128)).try_into().unwrap();
        Snowflake {
            id: Self::generate_id(timestamp, worker_id, process_id, increment),
            timestamp: timestamp,
            worker_id: worker_id,
            process_id: process_id,
            increment: increment,
        }
    }

    fn from_id(id: u64) -> Self {
        let timestamp = id >> 22;
        let worker_id = (id & 0x3e0000) >> 17;
        let process_id = (id & 0x1f000) >> 12;
        let increment = id & 0xfff;
        Snowflake {
            id: id,
            timestamp: timestamp,
            worker_id: worker_id,
            process_id: process_id,
            increment: increment,
        }
    }

    fn generate_id(timestamp: u64, worker_id: u64, process_id: u64, increment: u64) -> u64 {
        let mut id = timestamp << WORKER_BITS << PROCESS_BITS << INCREMENT_BITS;
        id |= (worker_id % (MAX_WORKER_ID as u64)) << PROCESS_BITS << INCREMENT_BITS;
        id |= (process_id % (MAX_PROCESS_ID as u64)) << INCREMENT_BITS;
        id |=  increment % MAX_INCREMENT_ID as u64;
        id
    }
}

impl From<&str> for Snowflake {
    fn from(s: &str) -> Self {
        Snowflake::new(s.parse::<u64>().unwrap())
    }
}

impl From<String> for Snowflake {
    fn from(s: String) -> Self {
        Snowflake::new(s.parse::<u64>().unwrap())
    }
}

impl Into<String> for Snowflake {
    fn into(self) -> String {
        self.id.to_string()
    }
}

File: src/Snowflake.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Snowflake = exports.MAX_TIMESTAMP = exports.MAX_INCREMENT_ID = exports.MAX_PROCESS_ID = exports.MAX_WORKER_ID = exports.INCREMENT_BITS = exports.PROCESS_ID_BITS = exports.WORKER_ID_BITS = exports.VIVA_EPOCH = void 0;
exports.VIVA_EPOCH = 1660769388000n;
exports.WORKER_ID_BITS = 5n;
exports.PROCESS_ID_BITS = 5n;
exports.INCREMENT_BITS = 12n;
exports.MAX_WORKER_ID = 1n << exports.WORKER_ID_BITS;
exports.MAX_PROCESS_ID = 1n << exports.PROCESS_ID_BITS;
exports.MAX_INCREMENT_ID = 1n << exports.INCREMENT_BITS;
exports.MAX_TIMESTAMP = 1n << (64n - exports.WORKER_ID_BITS - exports.PROCESS_ID_BITS - exports.INCREMENT_BITS);
class Snowflake {
    id;
    timestamp;
    workerId;
    datacenterId;
    incrementId;
    constructor(snowflake) {
        this.from_id(snowflake);
    }
    static from_string(snowflake) {
        return new Snowflake(BigInt(snowflake));
    }
    static from_id(id) {
        return new Snowflake(id);
    }
    static create(workerId, datacenterId, incrementId) {
        const timestamp = (BigInt(Date.now()) - exports.VIVA_EPOCH) % exports.MAX_TIMESTAMP;
        return new this(Snowflake.generate_id(timestamp, workerId, datacenterId, incrementId));
    }
    static generate_id(timestamp, workerId, datacenterId, incrementId) {
        let id = timestamp << exports.WORKER_ID_BITS << exports.PROCESS_ID_BITS << exports.INCREMENT_BITS;
        id |= (BigInt(workerId) % exports.MAX_WORKER_ID) << exports.PROCESS_ID_BITS << exports.INCREMENT_BITS;
        id |= (BigInt(datacenterId) % exports.MAX_PROCESS_ID) << exports.INCREMENT_BITS;
        id |= BigInt(incrementId) % exports.MAX_INCREMENT_ID;
        return id;
    }
    from_id(id) {
        if (id < 0) {
            throw new Error(`Invalid snowflake id ${id}`);
        }
        if ((id + 1n) === 1n) {
            throw new Error(`Invalid snowflake id ${id}`);
        }
        // timestamp
        const timestamp = id >> (exports.WORKER_ID_BITS + exports.PROCESS_ID_BITS + exports.INCREMENT_BITS);
        // worker id
        const workerId = (id >> (exports.PROCESS_ID_BITS + exports.INCREMENT_BITS)) & ((1n << exports.WORKER_ID_BITS) - 1n);
        // datacenter id
        const datacenterId = (id >> exports.INCREMENT_BITS) & ((1n << exports.PROCESS_ID_BITS) - 1n);
        // increment id
        const incrementId = id & ((1n << exports.INCREMENT_BITS) - 1n);
        this.id = id;
        this.timestamp = timestamp;
        this.workerId = workerId;
        this.datacenterId = datacenterId;
        this.incrementId = incrementId;
    }
    // getters for the snowflake parts
    get_timestamp() {
        return this.timestamp;
    }
    get_worker_id() {
        return this.workerId;
    }
    get_process_id() {
        return this.datacenterId;
    }
    get_increment_id() {
        return this.incrementId;
    }
    get_id() {
        return this.id;
    }
}
exports.Snowflake = Snowflake;
